tests/sentry/sentry_metrics/test_last_seen_updater.py

```diff
@@ -0,0 +1,227 @@
+        "not JSON",
+        encoding="utf-8",
+    )
+
+
+def headerless_kafka_payload(payload_bytes):
+    return KafkaPayload(key=bytes("fake-key", encoding="utf-8"), value=payload_bytes, headers=[])
+
+
+def kafka_message(kafka_payload):
+    return Message(
+        partition=Partition(Topic("fake-topic"), 1),
+        offset=1,
+        payload=kafka_payload,
+        timestamp=datetime.now(),
+    )
+
+
+def test_retrieve_db_read_keys_meta_field_present_with_db_keys():
+    message = kafka_message(headerless_kafka_payload(mixed_payload()))
+    key_set = retrieve_db_read_keys(message)
+    assert key_set == {2000, 2001, 2002}
+
+
+def test_retrieve_db_read_keys_meta_field_not_present():
+    message = kafka_message(headerless_kafka_payload(empty_payload()))
+    key_set = retrieve_db_read_keys(message)
+    assert key_set == set()
+
+
+def test_retrieve_db_read_keys_meta_field_bad_json():
+    message = kafka_message(headerless_kafka_payload(bad_payload()))
+    key_set = retrieve_db_read_keys(message)
+    assert key_set == set()
+
+
+class TestLastSeenUpdaterEndToEnd(TestCase):
+    @staticmethod
+    def processing_factory():
+        return _last_seen_updater_processing_factory(max_batch_time=1.0, max_batch_size=1)
+
+    def setUp(self):
+        self.org_id = 1234
+        self.stale_id = 2001
+        self.fresh_id = 2002
+        self.stale_last_seen = timezone.now() - timedelta(days=1)
+        self.fresh_last_seen = timezone.now() - timedelta(hours=1)
+        StringIndexer.objects.create(
+            organization_id=self.org_id,
+            string="e2e_0",
+            id=self.stale_id,
+            last_seen=self.stale_last_seen,
+        )
+        StringIndexer.objects.create(
+            organization_id=self.org_id,
+            string="e2e_1",
+            id=self.fresh_id,
+            last_seen=self.fresh_last_seen,
+        )
+
+    def tearDown(self):
+        StringIndexer.objects.filter(id=self.fresh_id).delete()
+        StringIndexer.objects.filter(id=self.stale_id).delete()
+
+    def test_basic_flow(self):
+        with override_options({"sentry-metrics.last-seen-updater.accept-rate": 1.0}):
+            # we can't use fixtures with unittest.TestCase
+            message = kafka_message(headerless_kafka_payload(mixed_payload()))
+            processing_strategy = self.processing_factory().create(lambda x: print("commit called"))
+            processing_strategy.submit(message)
+            processing_strategy.poll()
+            processing_strategy.join(1)
+
+        fresh_item = StringIndexer.objects.get(id=self.fresh_id)
+        assert fresh_item.last_seen == self.fresh_last_seen
+
+        stale_item = StringIndexer.objects.get(id=self.stale_id)
+        # without doing a bunch of mocking around time objects, stale_item.last_seen
+        # should be approximately equal to timezone.now() but they won't be perfectly equal
+        assert (timezone.now() - stale_item.last_seen) < timedelta(seconds=30)
```

Comment:
I would avoid hardcoding a number of seconds here. What about instead asserting that the the `last_seen` is higher than the one you initialized the row with? So you do not need to rely on a time interval.